internal/state/storage/storage.go

```diff
@@ -0,0 +1,293 @@
+	storage.lastOperationEndpointsMetric.Collect(metrics)
+	storage.lastOperationDurationMetric.Collect(metrics)
+}
+
+// Snapshotter is an interface for exporting and importing cluster state.
+type Snapshotter interface {
+	// Snapshot returns a snapshot of the current state.
+	Snapshot() *storagepb.StateSnapshot
+
+	// LoadFromSnapshot loads the state from the snapshot.
+	LoadFromSnapshot(snapshot *storagepb.StateSnapshot)
+}
+
+// New creates a new instance of Storage.
+func New(path string, state Snapshotter, logger *zap.Logger) *Storage {
+	return &Storage{
+		state:  state,
+		logger: logger.With(zap.String("component", "storage"), zap.String("path", path)),
+		path:   path,
+
+		operationsMetric: prom.NewCounterVec(prom.CounterOpts{
+			Name: "discovery_storage_operations_total",
+			Help: "The total number of storage operations.",
+		}, []string{labelOperation, labelStatus}),
+		lastSnapshotSizeMetric: prom.NewGaugeVec(prom.GaugeOpts{
+			Name: "discovery_storage_last_snapshot_size_bytes",
+			Help: "The size of the last processed snapshot in bytes.",
+		}, []string{labelOperation}),
+		lastOperationClustersMetric: prom.NewGaugeVec(prom.GaugeOpts{
+			Name: "discovery_storage_last_operation_clusters",
+			Help: "The number of clusters in the snapshot of the last operation.",
+		}, []string{labelOperation}),
+		lastOperationAffiliatesMetric: prom.NewGaugeVec(prom.GaugeOpts{
+			Name: "discovery_storage_last_operation_affiliates",
+			Help: "The number of affiliates in the snapshot of the last operation.",
+		}, []string{labelOperation}),
+		lastOperationEndpointsMetric: prom.NewGaugeVec(prom.GaugeOpts{
+			Name: "discovery_storage_last_operation_endpoints",
+			Help: "The number of endpoints in the snapshot of the last operation.",
+		}, []string{labelOperation}),
+		lastOperationDurationMetric: prom.NewGaugeVec(prom.GaugeOpts{
+			Name: "discovery_storage_last_operation_duration_seconds",
+			Help: "The duration of the last operation in seconds.",
+		}, []string{labelOperation}),
+	}
+}
+
+// Start starts the storage loop that periodically saves the state.
+func (storage *Storage) Start(ctx context.Context, clock clockwork.Clock, interval time.Duration) error {
+	storage.logger.Info("start storage loop", zap.Duration("interval", interval))
+
+	ticker := clock.NewTicker(interval)
+
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			storage.logger.Info("received shutdown signal")
+
+			if err := storage.Save(); err != nil {
+				return fmt.Errorf("failed to save state on shutdown: %w", err)
+			}
+
+			if errors.Is(ctx.Err(), context.Canceled) {
+				return nil
+			}
+
+			return ctx.Err()
+		case <-ticker.Chan():
+			if err := storage.Save(); err != nil {
+				storage.logger.Error("failed to save state", zap.Error(err))
+			}
+		}
+	}
+}
+
+// Save saves all clusters' states into the persistent storage.
+func (storage *Storage) Save() (err error) {
+	if storage.inProgress {
```

Comment:
I'm confused about this flag: it looks like this Save runs in the same goroutine and it blocks until it's done.
Why do we need this flag then?

On other hand if it can run in different goroutines then we should have thread safety here.